### DISCUSSION TITLE
Slow autocomplete for ssh config host names

### DISCUSSION THREAD

**Discussion by laurensV** [microsoft/WSL#1999] [General]
Autocompletion for server names specified in ~/.ssh/config works, but is slow. It takes a few seconds to autocomplete the server names. Other autocomplete functionality is working instantly, only the ssh config autocomplete functionaltity is very slow. On native linux this is working instantly.

My windows build: 10.0.15046 (however, I have had this on all builds so far)

**Comment by sigv** [comment]
For what it's worth, on my box, it seems _bash_completion_ is set to complete for machines that are reporting their uptime but the package which actually manages that was not installed by default.

Ref from _/usr/share/bash-completion/bash_completion_:

```
    # Add hosts reported by ruptime.
    COMPREPLY+=( $( compgen -W \
        "$( ruptime 2>/dev/null | awk '!/^ruptime:/ { print $1 }' )" \
        -- "$cur" ) )
```

The package _rwho_ package provides `ruptime` so you can just `sudo apt install rwho` to get that binary.
Or you can just comment out the part of the script that deals with polling `ruptime`.

**Comment by Oylex** [comment]
I'm having the same problem, I've tried installing rwho, but it says the package doesn't exists.

So I've commented out the code from sigv's post and it now works like a charm!

**Comment by jasonrohrer** [comment]
Installing rwho worked great on Ubuntu Xenial.

Not sure that this issue is specific to Microsoft WSL....

I've been suffering from this problem for more than a year, and I just finally got frustrated enough to look for a solution by searching for "ssh autocomplete slow" and found this.

Thank you!

**Comment by therealkenc** [comment]
> Not sure that this issue is specific to Microsoft WSL....

Possibly. Performance (if that is the word) seems alright on the bash-completion that ships with 20.04 on recent WSL1, _without_ having installed `rwho`.

![ssh-completion](https://user-images.githubusercontent.com/7049700/93970170-6432fd00-fd22-11ea-97e4-c0145f1b3e05.gif)

Someone inside the intersection of (a) able to reproduce the behavior and (b) highly motivated, could do an `strace -tt -f -o bash.strace /bin/bash` and see where `bash` is spending its time. But whatever it is doing, it is (at the end of the day) going to take whatever time it takes.
